// uniforms/lib/bridges/Bridge.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Bridge = function () {
    function Bridge(schema) {
        _classCallCheck(this, Bridge);

        this.schema = schema;
    }

    // Check, if this bridge is compatibile with given schema


    _createClass(Bridge, [{
        key: "getError",


        // Field's scoped error
        value: function getError(name, error) {
            return this.schema.getError(name, error);
        }

        // Field's scoped error message

    }, {
        key: "getErrorMessage",
        value: function getErrorMessage(name, error) {
            return this.schema.getErrorMessage(name, error);
        }

        // All error messages from error

    }, {
        key: "getErrorMessages",
        value: function getErrorMessages(error) {
            return this.schema.getErrorMessages(error);
        }

        // Field's definition (`field` prop)

    }, {
        key: "getField",
        value: function getField(name) {
            return this.schema.getField(name);
        }

        // Field's initial value

    }, {
        key: "getInitialValue",
        value: function getInitialValue(name, props) {
            return this.schema.getInitialValue(name, props);
        }

        // Field's props

    }, {
        key: "getProps",
        value: function getProps(name) {
            return this.schema.getProps(name);
        }

        // Field's subfields (or first-level fields)

    }, {
        key: "getSubfields",
        value: function getSubfields(name) {
            return this.schema.getSubfields(name);
        }

        // Field's type (ex. Number, String)

    }, {
        key: "getType",
        value: function getType(name) {
            return this.schema.getType(name);
        }

        // Function with one argument - model - which throws errors
        // when model is invalid

    }, {
        key: "getValidator",
        value: function getValidator(options) {
            return this.schema.getValidator(options);
        }
    }], [{
        key: "check",
        value: function check(schema) {
            return schema && schema.getError && schema.getErrorMessage && schema.getErrorMessages && schema.getField && schema.getInitialValue && schema.getProps && schema.getSubfields && schema.getType && schema.getValidator;
        }
    }]);

    return Bridge;
}();

exports.default = Bridge;
