// uniforms-bootstrap3/lib/components/fields/BoolField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _uniforms = require('uniforms');

var _FormGroup = require('./FormGroup');

var _FormGroup2 = _interopRequireDefault(_FormGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Bool = function Bool(props) {
    return _react2.default.createElement(
        _FormGroup2.default,
        props,
        _react2.default.createElement(
            'section',
            { className: (0, _classnames2.default)(props.inputClassName, 'checkbox' + (props.inline ? '-inline' : '')) },
            _react2.default.createElement(
                'label',
                { onClick: function onClick() {
                        return props.onChange(!props.value);
                    } },
                _react2.default.createElement('input', {
                    checked: props.value,
                    className: 'hidden',
                    disabled: props.disabled,
                    name: props.name,
                    onChange: function onChange() {
                        return props.onChange(!props.value);
                    },
                    type: 'checkbox'
                }),
                props.label
            )
        )
    );
};

exports.default = (0, _uniforms.connectField)(Bool);
