// uniforms-bootstrap3/lib/components/fields/FormGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _gridClassName = require('../../lib/gridClassName');

var _gridClassName2 = _interopRequireDefault(_gridClassName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var makeHelp = function makeHelp(help, helpClassName) {
    return help && _react2.default.createElement(
        'span',
        { className: helpClassName || 'text-muted' },
        help
    );
};

var FormGroup = function FormGroup(_ref) {
    var children = _ref.children;
    var className = _ref.className;
    var // class name for the whole .form-group
    disabled = _ref.disabled;
    var // boolean, if true, show fields as disabled
    error = _ref.error;
    var // error validation response
    grid = _ref.grid;
    var // grid is either a int [1-11] or object {xs:6,sm:4,md:2}
    help = _ref.help;
    var // help text
    helpClassName = _ref.helpClassName;
    var // class name for the help text (default: 'text-muted')
    label = _ref.label;
    var // string label (or false)
    required = _ref.required;
    var feedbackable = _ref.feedbackable;
    var // only some input types support feedback icons
    wrapClassName // class name for the section wrapping the input(s)
    = _ref.wrapClassName;
    return _react2.default.createElement(
        'section',
        {
            className: (0, _classnames2.default)(className, 'field', 'form-group', {
                'has-feedback': error && feedbackable,
                'has-error': error,
                disabled: disabled,
                required: required
            })
        },
        label && _react2.default.createElement(
            'label',
            { className: (0, _classnames2.default)('control-label', (0, _gridClassName2.default)(grid, 'label')) },
            label
        ),
        (grid || wrapClassName) && _react2.default.createElement(
            'section',
            { className: (0, _classnames2.default)(wrapClassName, (0, _gridClassName2.default)(grid, 'input')) },
            children,
            makeHelp(help, helpClassName)
        ),
        !grid && !wrapClassName && children,
        !grid && !wrapClassName && makeHelp(help, helpClassName)
    );
};

exports.default = FormGroup;
